primitives/core/src/bandersnatch.rs

```diff
@@ -0,0 +1,1010 @@
+		/// transform part of the scheme. If unsure just give it a unique label
+		/// depending on the actual usage of the signing data
+		/// (@burges: or leave it empty? There is already the `label` field for contextualization).
+		/// The `vrf_inputs` is a sequence of [`VrfInput`]s to be signed and which
+		/// contribute to the actual output bytes produced via the VRF.
+		pub fn new<T: Into<VrfIosVec<VrfInput>>>(
+			label: &'static [u8],
+			transcript_data: &[&[u8]],
+			vrf_inputs: T,
+		) -> Self {
+			let mut transcript = Transcript::new_labeled(label);
+			transcript_data.iter().for_each(|data| transcript.append_slice(data));
+			VrfSignData { transcript, vrf_inputs: vrf_inputs.into() }
+		}
+
+		/// Construct a new data to be signed from an iterator of [`VrfInput`]s.
+		///
+		/// Fails if the `vrf_inputs` yields more elements than [`MAX_VRF_IOS`]
+		// TODO @davxy: maybe we can just provide one constructor...
+		pub fn from_iter<T: IntoIterator<Item = VrfInput>>(
+			label: &'static [u8],
+			transcript_data: &[&[u8]],
+			vrf_inputs: T,
+		) -> Result<Self, ()> {
+			let vrf_inputs: Vec<VrfInput> = vrf_inputs.into_iter().collect();
+			let bounded = VrfIosVec::try_from(vrf_inputs).map_err(|_| ())?;
+			Ok(Self::new(label, transcript_data, bounded))
+		}
+
+		/// Appends a message to the transcript
+		pub fn push_transcript_data(&mut self, data: &[u8]) {
+			self.transcript.append_slice(data);
+		}
+
+		/// Appends a [`VrfInput`] to the vrf inputs to be signed.
+		///
+		/// On failure, gives back the [`VrfInput`] parameter.
+		pub fn push_vrf_input(&mut self, vrf_input: VrfInput) -> Result<(), VrfInput> {
+			self.vrf_inputs.try_push(vrf_input)
+		}
+
+		/// Create challenge from input transcript within the signing data.
+		pub fn challenge<const N: usize>(&self) -> [u8; N] {
+			let mut output = [0; N];
+			let mut t = self.transcript.clone();
+			let mut reader = t.challenge(b"Prehashed for Ed25519");
+			reader.read_bytes(&mut output);
+			output
+		}
+	}
+
+	/// VRF signature.
+	#[derive(Clone, Debug, PartialEq, Eq, Encode, Decode, MaxEncodedLen, TypeInfo)]
+	pub struct VrfSignature {
+		/// VRF (pre)outputs.
+		pub vrf_outputs: VrfIosVec<VrfOutput>,
+		/// VRF signature.
+		pub signature: Signature,
+	}
+
+	#[cfg(feature = "full_crypto")]
+	impl VrfCrypto for Pair {
+		type VrfInput = VrfInput;
+		type VrfOutput = VrfOutput;
+		type VrfSignData = VrfSignData;
+		type VrfSignature = VrfSignature;
+	}
+
+	#[cfg(feature = "full_crypto")]
+	impl VrfSecret for Pair {
+		fn vrf_sign(&self, data: &Self::VrfSignData) -> Self::VrfSignature {
+			// TODO: @davxy
+			// Hack used because backend signature type is generic over the number of ios
+			// @burdges can we provide a Vec version in `bandersnatch_vrfs` crate?
+			match data.vrf_inputs.len() {
+				0 => self.vrf_sign_gen::<0>(data),
+				1 => self.vrf_sign_gen::<1>(data),
+				2 => self.vrf_sign_gen::<2>(data),
+				3 => self.vrf_sign_gen::<3>(data),
+				_ => unreachable!(),
```

Comment:
Unfortunately I had to resort to this workaround.

Can we provide a Vec version of Signature directly in `bandersnatch_vrfs` crate?